File: android/src/main/java/com/capacitorjs/plugins/call/CallPlugin.java
package com.capacitorjs.plugins.call;

import android.content.Intent;

import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "Call")
public class CallPlugin extends Plugin {

    private final Call implementation = new Call();

    @PluginMethod
    public void call(PluginCall call) {
        String number = call.getString("number");

        if (number == null || number.isEmpty()) {
            call.reject("error-missing-number");
            return;
        }

        try {
            Intent intent = implementation.call(number);
            if (intent == null) {
                call.reject("error-call-failed");
            } else {
                this.getContext().startActivity(intent);
            }
        } catch (Exception e) {
            call.reject(e.getMessage());
        }
    }
}

File: ios/Sources/CallPlugin/CallPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(CallPlugin)
public class CallPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "CallPlugin"
    public let jsName = "Call"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "call", returnType: CAPPluginReturnPromise),
    ]

    @objc func call(_ call: CAPPluginCall) {
        guard var formattedNumber = call.getString("number")
        else {
            call.reject("error-missing-number")
            return
        }
        
        if !formattedNumber.starts(with:"tel:") {
            formattedNumber = "tel:" + formattedNumber
        }
        
        formattedNumber = formattedNumber.replacingOccurrences(of: "#", with: "%23")
        guard let number = URL(string:formattedNumber)
        else {
            call.reject("error-call-failed")
            return
        }
        DispatchQueue.main.async {
            UIApplication.shared.open(number, options: [:]) { success in
                if success {
                    call.resolve()
                } else {
                    call.reject("error-call-failed")
                }
            }
        }
    }
}
